binding.cc

```diff
@@ -600,8 +620,59 @@ struct Database {
 
 private:
   uint32_t priorityWork_;
+  std::string location_;
+
+  // for separation of concerns the threadsafe functionality was kept at the global
+  // level and made a friend so it is explict where the threadsafe boundary exists
+  friend leveldb::Status threadsafe_open(const leveldb::Options &options,
+                                         bool allow_multi_threading,
+                                         Database &db_instance);
+  friend leveldb::Status threadsafe_close(Database &db_instance);
 };
 
+
+leveldb::Status threadsafe_open(const leveldb::Options &options,
+                                bool allow_multi_threading,
+                                Database &db_instance) {
+  std::unique_lock<std::mutex> lock(handles_mutex);
+
+  auto it = db_handles.find(db_instance.location_);
+  if (it != db_handles.end() && allow_multi_threading) {
+    // Database already opened for this location
+    ++(it->second.open_handle_count);
+    db_instance.db_ = it->second.db;
+  } else {
+    // Database not opened yet for this location
+    LevelDbHandle handle = {nullptr, 0};
+    leveldb::Status status = leveldb::DB::Open(options, db_instance.location_, &handle.db);
+    if (!status.ok()) {
+      return status;
+    }
+    handle.open_handle_count++;
+    db_instance.db_ = handle.db;
+    db_handles[db_instance.location_] = handle;
+  }
+
+  return leveldb::Status::OK();
+}
+
+leveldb::Status threadsafe_close(Database &db_instance) {
+  std::unique_lock<std::mutex> lock(handles_mutex);
+  db_instance.db_ = NULL;  // ensure db_ pointer is nullified in Database instance
+
+  auto it = db_handles.find(db_instance.location_);
+  if (it == db_handles.end()) {
+    return leveldb::Status::NotFound("Database handle not found for the given location");
```

Comment:
An inability to close is a bad, inescapable situation, so I prefer not to push that problem to userland. We should either change this to an `assert` (if we're sure it never happens) or just silently return `Status::OK()`.

Reply:
Updated.  I went with the `OK` route and added a note for those in the future maintainers